### PR TITLE
Fix mousemove bug, ghost offset bug, and many others

### PR DESCRIPTION
**Pull request by ntilwalli**
**Fix mousemove bug:** previously double-clicking an element left the mousemove stream active.
**Ghost offset bug:** The ghost was previously being shifted to the right and down in a non-intuitive way, now the ghost starts in the original place, 
**Bug requiring children if no handle option set:**  previously having an empty list and no `handle` option set caused a crash.  Now, the list does not react until there are two or more children
**Remove need for emitBetween**:  Use `endWhen`
**Use border-box:**  Added border-box to simple and horizontal examples to get rid of ghost rendering bigger than original element when clicked.
**Use relative position for parent css selectors:** This lib requires the parent elements to have a position set to anything other than `static`.  It can't be ensured in the code directly since we may overwrite the existing position class.  relative positioning has been added to simple and horizontal examples.
**Add code to implement `ghostClass` option:** this was not previously implemented, though the option existed in the `options`

### CODE REVIEW

File: examples/simple/css/main.css [hunk 1]
```diff
@@ -2,6 +2,13 @@
     font-size: 35px;
 }
 
+html {
+  box-sizing: border-box;
```
Comment (jvanbruegge) [inline examples/simple/css/main.css]: can you indent with 4 spaces please?

File: examples/simple/src/index.ts [hunk 2]
```diff
@@ -24,7 +24,7 @@ function main({ DOM } : Sources) : Sinks
             li('.li', '', ['Option 6']),
         ])
     )
-    .compose(makeSortable<Stream<VNode>>(DOM));
+    .compose(makeSortable<Stream<VNode>>(DOM, {ghostClass: '.ghost'}));
```
Comment (jvanbruegge) [inline examples/simple/src/index.ts]: Please add spaces inside the braces

File: src/eventHandlers/mousedown.ts [hunk 3]
```diff
@@ -1,33 +1,39 @@
 import { VNode } from '@cycle/dom';
 import { select } from 'snabbdom-selector';
-import { EventHandler, MouseOffset } from '../definitions';
+import { EventHandler, MouseOffset, SortableOptions} from '../definitions';
```
Comment (jvanbruegge) [inline src/eventHandlers/mousedown.ts]: Please add a space inside the brace

**Comment by ntilwalli** [comment]
This PR fixes https://github.com/cyclejs-community/cyclejs-sortable/issues/14

File: src/eventHandlers/mousedown.ts [hunk 4]
```diff
@@ -1,33 +1,39 @@
...
 
-import { getIndex, getGhostStyle, findParent, addAttributes, replaceNode, getBodyStyle, addKeys } from '../helpers';
+import { getIndex, getGhostStyle, findParent, addAttributes, addGhostClass, replaceNode, getBodyStyle, addKeys } from '../helpers';
 
 /**
  * Used to create the ghost and hide the item dragged
  * @type {EventHandler}
  */
-export const mousedownHandler : EventHandler = (node, event, options) => {
+export const mousedownHandler : EventHandler = (node, event, options: SortableOptions) => {
```
Comment (jvanbruegge) [inline src/eventHandlers/mousedown.ts]: The type of options is already defined in `EventHandler`

File: src/eventHandlers/mousedown.ts [hunk 5]
```diff
@@ -1,33 +1,39 @@
...
     const item : Element = findParent(event.target as Element, options.parentSelector + ' > *');
     const itemRect : ClientRect = item.getBoundingClientRect();
-    const mouseOffset : MouseOffset = {
+    const parentNode: Element = item.parentElement;
+    const parentRect: ClientRect = parentNode.getBoundingClientRect();
+    const mouseOffset = {
```
Comment (jvanbruegge) [inline src/eventHandlers/mousedown.ts]: Please leave the type in

File: src/eventHandlers/mousedown.ts [hunk 6]
```diff
@@ -1,33 +1,39 @@
...
+        parentTop: parentRect.top,
     };
 
     const body : Element = findParent(event.target as Element, 'body');
     body.setAttribute('style', getBodyStyle());
 
     const parent : VNode = select(options.parentSelector, node)[0];
     const index : number = getIndex(item);
-
+    const style = getGhostStyle(event, mouseOffset, item)
```
Comment (jvanbruegge) [inline src/eventHandlers/mousedown.ts]: We are not reusing `style`, so no need to declare a var

File: src/eventHandlers/mousemove.ts [hunk 7]
```diff
@@ -32,9 +32,9 @@ export const mousemoveHandler : EventHandler = (node, event, options) => {
 
     const newIndex : number = maxIntersection[1] === itemIndex ? itemIndex :
         (-itemIntersection > maxArea - itemArea ? maxIntersection[1] : itemIndex);
-
+    const style = updateGhostStyle(event, mouseOffset, ghost.elm as Element)
```
Comment (jvanbruegge) [inline src/eventHandlers/mousemove.ts]: Same here

File: src/helpers.ts [hunk 8]
```diff
@@ -100,24 +100,27 @@ export function replaceNode(root : VNode, selector : string, replacement : VNode
  * @return {string} the style value
  */
 export function getGhostStyle(event : MouseEvent, mouseOffset : MouseOffset, item : Element) : string
-{
+{
```
Comment (jvanbruegge) [inline src/helpers.ts]: Trailing spaces

File: src/helpers.ts [hunk 9]
```diff
@@ -100,24 +100,27 @@ export function replaceNode(root : VNode, selector : string, replacement : VNode
...
  */
 export function getGhostStyle(event : MouseEvent, mouseOffset : MouseOffset, item : Element) : string
-{
+{   
     const itemRect : ClientRect = item.getBoundingClientRect();
-    return 'z-index: 5; margin: 0; pointer-events: none; position: absolute; width: '
-        + itemRect.width + 'px; ' + 'height: ' + itemRect.height + 'px; top: '
-        + (event.clientY + mouseOffset.y + window.screenY) + 'px; left: '
-        + (event.clientX + mouseOffset.x + window.scrollX) + 'px;';
+    const out = 'z-index: 5; margin: 0; pointer-events: none; position: absolute; ' +
```
Comment (jvanbruegge) [inline src/helpers.ts]: no need to create var here

File: src/helpers.ts [hunk 10]
```diff
@@ -100,24 +100,27 @@ export function replaceNode(root : VNode, selector : string, replacement : VNode
...
+        'top: ' + (mouseOffset.itemTop - mouseOffset.parentTop) + 'px; ' +
+        'left: ' + (mouseOffset.itemLeft - mouseOffset.parentLeft) + 'px;';
+    return out;
 }
 
 /**
  * Returns the updated style for this ghost element
  */
-export function updateGhostStyle(event : MouseEvent, mouseOffset : MouseOffset, ghost : Element) : string
+export function updateGhostStyle(event : any, mouseOffset : MouseOffset, ghost : Element) : string
```
Comment (jvanbruegge) [inline src/helpers.ts]: Is this needed for TouchEvents? If so I would use a sum type

File: src/helpers.ts [hunk 11]
```diff
@@ -100,24 +100,27 @@ export function replaceNode(root : VNode, selector : string, replacement : VNode
...
 {
     const prevStyle : string = ghost.getAttribute('style');
     
-    return prevStyle.substring(0, prevStyle.indexOf(' top:')) + ' top: '
-        + (event.clientY + mouseOffset.y + window.scrollY) + 'px; left: '
-        + (event.clientX + mouseOffset.x + window.scrollX) + 'px;';
+    const out = prevStyle.substring(0, prevStyle.indexOf(' top:')) + ' top: '
```
Comment (jvanbruegge) [inline src/helpers.ts]: No var needed

File: src/helpers.ts [hunk 12]
```diff
@@ -148,14 +151,33 @@ export function findParent(node : Element, selector : string) : Element
...
- * Adds the given attribute savely to the VNode
+ * Adds the given attribute safely to the VNode
  * @param {VNode} node the VNode to add the attributes on
  * @param {any} addition the new attributes
  * @return {VNode} the newly created VNode
  */
-export function addAttributes(e : VNode, newAttr : { [attr : string]: any }) : VNode
+export function addAttributes(e : VNode, newAttr : { [attr : string]: any }, ghostClass?: string) : VNode
+{
+
```
Comment (jvanbruegge) [inline src/helpers.ts]: Please remove line

File: src/makeSortable.ts [hunk 13]
```diff
@@ -17,6 +16,19 @@ function augmentEvent(ev : any) : any {
     return ev;
 }
 
+function augmentStartDistance(ev : any, startX: any, startY: any) : any {
```
Comment (jvanbruegge) [inline src/makeSortable.ts]: Please add spaces before colons

Comment (jvanbruegge) [reply src/makeSortable.ts]: I'm not sure if we should modify the event. Have to think about this

File: src/makeSortable.ts [hunk 14]
```diff
@@ -17,6 +16,19 @@ function augmentEvent(ev : any) : any {
...
+    ev.distX = ev.clientX - startX
+    ev.distY = ev.clientY - startY
+    return ev;
+}
+
+function moreThanOneChild(node: VNode) {
+    return !node.children || node.children.length > 1
```
Comment (jvanbruegge) [inline src/makeSortable.ts]: Semicolon

**Review by jvanbruegge — CHANGES_REQUESTED** [review]
I just had a quick look over it, thanks for the PR. I have to check it more thoroughly, this are mostly style comments

**Comment by ntilwalli** [comment]
Updated

**Comment by ntilwalli** [comment]
Performance can get laggy on larger lists.  What can we do to improve perf a bit?

**Comment by jvanbruegge** [comment]
I'm sorry that I forgot this PR.
In the current state I cannot merge this. Please split the PR in parts (one oer bug/feature) and please use tslint to check for linting errors

**Comment by ntilwalli** [comment]
@jvanbruegge is it possible to avoid breaking this up into separate PRs?  There are a lot of bug fixes and unwinding all of them will take a while.  Some fixes are dependent on others...  Can we discuss the changes instead?  And work through them?

**Comment by jvanbruegge** [comment]
I've resolved the merge conflict in #16 
But it seems like all examples stop working with this

**Pull request closed without merging**